File: go/sortedSquares.go
package sortedsquares

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func sq(x int) int {
	return x * x
}

func runSortedSquares__tail(input []int, accum []int) []int {
	if len(input) == 0 {
		return accum
	}

	first := input[0]
	lastIndex := len(input) - 1
	last := input[lastIndex]

	if abs(first) > last {
		return runSortedSquares__tail(input[1:], append([]int{sq(first)}, accum...))
	}

	return runSortedSquares__tail(input[0:lastIndex], append([]int{sq(last)}, accum...))
}

func runSortedSquares__goto(input []int, accum []int) []int {
Entry:

	if len(input) == 0 {
		return accum
	}

	first := input[0]
	lastIndex := len(input) - 1
	last := input[lastIndex]

	if abs(first) > last {
		accum = append([]int{sq(first)}, accum...)
		input = input[1:]
		goto Entry
	}

	accum = append([]int{sq(last)}, accum...)
	input = input[0:lastIndex]
	goto Entry
}

// SortedSquares__tail creates a sorted list of squares from the input
func SortedSquares__tail(input []int) []int {
	return runSortedSquares__tail(input, []int{})
}

// SortedSquares__tail creates a sorted list of squares from the input
func SortedSquares__goto(input []int) []int {
	return runSortedSquares__goto(input, []int{})
}

// SortedSquares__recursive creates a sorted list of squares from the input
func SortedSquares__recursive(input []int) []int {
	if len(input) == 0 {
		return []int{}
	}

	first := input[0]
	lastIndex := len(input) - 1
	last := input[lastIndex]

	if abs(first) > last {
		return append(
			SortedSquares__recursive(input[1:]),
			sq(first),
		)
	}

	return append(
		SortedSquares__recursive(input[0:lastIndex]),
		sq(last),
	)
}

File: c/sorted-squares.c
#include "./int-vector.h"
#include <stdio.h>

int abs(int num) { return num < 0 ? -num : num; }

vector *sortedSquaresRecursive(vector *nums, vector *accum) {
	// Given an empty input.
	if (nums->size == 0) {
		return accum;
	}

	// Only one element remains. Add it and return the result.
	if (nums->size == 1) {
		vector_unshift(accum, nums->data[0] * nums->data[0]);
		return accum;
	}

	int first = nums->data[0];
	int last = nums->data[nums->size - 1];

	if (abs(first) > last) {
		vector_unshift(accum, first * first);
		vector_shift(nums);
	} else {
		vector_unshift(accum, last * last);
		vector_pop(nums);
	}

	return sortedSquaresRecursive(nums, accum);
}

vector *sortedSquares(vector *nums) {
	vector *initialAccumulator = vector_create_empty();
	vector *result = sortedSquaresRecursive(nums, initialAccumulator);

	vector_free(initialAccumulator);

	return result;
}

int main(int argc, char **argv) {
	int data[] = {-3, 1, 2, 5, 9};
	vector *nums = vector_create(5, data);
	vector *result = sortedSquares(nums);

	vector_print(result);
}
